File: src/main/java/com/gitlab/daring/image/command/env/EnvCommands.java
package com.gitlab.daring.image.command.env;

import com.gitlab.daring.image.command.Command;
import com.gitlab.daring.image.command.CommandRegistry;
import com.gitlab.daring.image.command.SimpleCommand;

import static org.bytedeco.javacpp.opencv_imgcodecs.imwrite;

public class EnvCommands {

    public static void register(CommandRegistry r) {
        EnvCommands f = new EnvCommands();
        r.register("write", f::newWriteCommand);
        r.register("get", f::newGetCommand);
        r.register("put", f::newPutCommand);
        r.register("read", ReadCommand::new);
        r.register("show", ShowCommand::new);
        r.register("task", SetTaskCommand::new);
    }

    public Command newWriteCommand(String... ps) {
        SimpleCommand c = new SimpleCommand(ps);
        String file = c.arg(0, "");
        String key = c.arg(1, "");
        return env -> imwrite(env.eval(file), env.getMat(key));
    }

    public Command newGetCommand(String... ps) {
        return env -> env.getMat(ps[0]).copyTo(env.mat);
    }

    public Command newPutCommand(String... ps) {
        return env -> env.putMat(ps[0], env.mat);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/env/EnvCommands.kt
package com.gitlab.daring.image.command.env

import com.gitlab.daring.image.command.Command
import com.gitlab.daring.image.command.CommandRegistry
import com.gitlab.daring.image.command.CommandUtils.newEnvCommand
import com.gitlab.daring.image.command.SimpleCommand

import org.bytedeco.javacpp.opencv_imgcodecs.imwrite

object EnvCommands {

    fun register(r: CommandRegistry) {
        r.register("write", this::newWriteCommand)
        r.register("get", this::newGetCommand)
        r.register("put", this::newPutCommand)
        r.register("read", ::ReadCommand)
        r.register("show", ::ShowCommand)
        r.register("task", ::SetTaskCommand)
    }

    fun newWriteCommand(args: Array<String>): Command {
        val c = SimpleCommand(args)
        val file = c.arg(0, "")
        val key = c.arg(1, "")
        c.func = { env -> imwrite(env.eval(file), env.getMat(key)) }
        return c
    }

    fun newGetCommand(args: Array<String>): Command {
        return newEnvCommand { env -> env.getMat(args[0]).copyTo(env.mat) }
    }

    fun newPutCommand(args: Array<String>): Command {
        return newEnvCommand { env -> env.putMat(args[0], env.mat) }
    }

}
